File: RadixCyclicNN/go/radixnet/chatgpt.go
package radixnet

// ChatGPT: the hosted alternative to a local model, over OpenAI's API
// (https://platform.openai.com).  The same three calls the tutor needs - the
// models a key may use (GET /models), a health check and one non-streaming
// completion (POST /chat/completions) - so a ChatGPTClient stands in for an
// OllamaClient wherever an LLMClient is taken.
//
// Configuration comes from the environment, the way the OpenAI tools expect
// it: OPENAI_API_KEY (or OPENAI_API_KEY_FILE, a file holding it, for Docker
// secrets), OPENAI_BASE_URL (any OpenAI-compatible server works), and
// RADIXNET_OPENAI_MODEL for the default model.  The key is read per request
// and never stored in a config, a record or a log line; unlike the local
// provider every prompt is sent to OpenAI and billed.

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"regexp"
	"strings"
	"time"
)

// DefaultChatGPTTimeout is how long one answer may take.
const DefaultChatGPTTimeout = 120 * time.Second

// DefaultChatGPTModelName is used when neither the caller nor the environment names a model.
const DefaultChatGPTModelName = "gpt-4o-mini"

// OpenAIURL is the hosted API; $OPENAI_BASE_URL overrides it.
const OpenAIURL = "https://api.openai.com/v1"

// ChatGPTError is an OpenAI API that cannot be reached, refuses the request, or answers something unusable.
type ChatGPTError struct {
	Message string
	Status  int
	Code    string
	Param   string
}

func (e *ChatGPTError) Error() string { return e.Message }

func chatgptErrorf(format string, args ...any) error {
	return &ChatGPTError{Message: fmt.Sprintf(format, args...)}
}

// IsChatGPTError reports whether err came from talking to the OpenAI API.
func IsChatGPTError(err error) bool {
	var target *ChatGPTError
	return errors.As(err, &target)
}

var localHosts = map[string]bool{
	"localhost": true, "127.0.0.1": true, "::1": true, "0.0.0.0": true, "host.docker.internal": true,
}

func isLocalHost(host string) bool {
	host = strings.ToLower(host)
	return localHosts[host] || strings.HasSuffix(host, ".local")
}

func envFlag(name string) bool {
	switch strings.ToLower(strings.TrimSpace(os.Getenv(name))) {
	case "1", "true", "yes", "on":
		return true
	}
	return false
}

// NormaliseChatGPTURL turns "host" or a full URL into a base URL without a trailing
// slash ("/v1" added when there is no path).  A key must not travel unencrypted to a
// remote host, so plain http:// is refused there unless RADIXNET_OPENAI_ALLOW_INSECURE is set.
func NormaliseChatGPTURL(raw string) (string, error) {
	text := strings.TrimSpace(raw)
	if text == "" {
		return "", fmt.Errorf("the OpenAI base URL is empty")
	}
	if !strings.Contains(text, "://") {
		text = "https://" + text
	}
	parsed, err := url.Parse(text)
	if err != nil {
		return "", fmt.Errorf("the OpenAI base URL is not a URL (got %q): %v", raw, err)
	}
	if parsed.Scheme != "http" && parsed.Scheme != "https" {
		return "", fmt.Errorf("the OpenAI base URL must be http:// or https:// (got %q)", raw)
	}
	if parsed.Host == "" {
		return "", fmt.Errorf("the OpenAI base URL has no host (got %q)", raw)
	}
	if parsed.Scheme == "http" && !isLocalHost(parsed.Hostname()) && !envFlag("RADIXNET_OPENAI_ALLOW_INSECURE") {
		return "", fmt.Errorf("refusing to send the API key unencrypted to %q: use https://, a local server, "+
			"or set RADIXNET_OPENAI_ALLOW_INSECURE=1", parsed.Hostname())
	}
	path := strings.TrimRight(parsed.Path, "/")
	if path == "" {
		path = "/v1"
	}
	return parsed.Scheme + "://" + parsed.Host + path, nil
}

// DefaultChatGPTURL is $OPENAI_BASE_URL (or $OPENAI_API_BASE), else the hosted API.
func DefaultChatGPTURL() string {
	configured := strings.TrimSpace(os.Getenv("OPENAI_BASE_URL"))
	if configured == "" {
		configured = strings.TrimSpace(os.Getenv("OPENAI_API_BASE"))
	}
	if configured == "" {
		return OpenAIURL
	}
	if normalised, err := NormaliseChatGPTURL(configured); err == nil {
		return normalised
	}
	return OpenAIURL
}

// DefaultChatGPTModel is $RADIXNET_OPENAI_MODEL, else gpt-4o-mini.
func DefaultChatGPTModel() string {
	if name := strings.TrimSpace(os.Getenv("RADIXNET_OPENAI_MODEL")); name != "" {
		return name
	}
	return DefaultChatGPTModelName
}

// ChatGPTAPIKey is $OPENAI_API_KEY, else the content of the file named by $OPENAI_API_KEY_FILE.
func ChatGPTAPIKey() string {
	if key := strings.TrimSpace(os.Getenv("OPENAI_API_KEY")); key != "" {
		return key
	}
	if path := strings.TrimSpace(os.Getenv("OPENAI_API_KEY_FILE")); path != "" {
		if raw, err := os.ReadFile(path); err == nil {
			return strings.TrimSpace(string(raw))
		}
	}
	return ""
}

// ChatGPTConfigured reports whether a key is available (its value is never returned or logged).
func ChatGPTConfigured() bool { return ChatGPTAPIKey() != "" }

// ChatGPTClient talks to one OpenAI-compatible endpoint.
type ChatGPTClient struct {
	URL     string
	Model   string
	Timeout time.Duration
	apiKey  string // empty: read from the environment per request
	client  *http.Client
}

// NewChatGPTClient normalises the URL and fills in the defaults for an empty url / model / timeout.
func NewChatGPTClient(rawURL, model string, timeout time.Duration) (*ChatGPTClient, error) {
	if strings.TrimSpace(rawURL) == "" {
		rawURL = DefaultChatGPTURL()
	}
	normalised, err := NormaliseChatGPTURL(rawURL)
	if err != nil {
		return nil, err
	}
	if strings.TrimSpace(model) == "" {
		model = DefaultChatGPTModel()
	}
	if timeout <= 0 {
		timeout = DefaultChatGPTTimeout
	}
	return &ChatGPTClient{URL: normalised, Model: strings.TrimSpace(model), Timeout: timeout, client: &http.Client{}}, nil
}

// Provider is "chatgpt".
func (c *ChatGPTClient) Provider() string { return ProviderChatGPT }

// BaseURL is the endpoint it talks to.
func (c *ChatGPTClient) BaseURL() string { return c.URL }

// ModelName is the model it answers with by default.
func (c *ChatGPTClient) ModelName() string { return c.Model }

// Configured reports whether a key is available for this client.
func (c *ChatGPTClient) Configured() bool { return c.key() != "" }

// String never shows the key.
func (c *ChatGPTClient) String() string { return fmt.Sprintf("ChatGPTClient(%s, %s)", c.URL, c.Model) }

func (c *ChatGPTClient) key() string {
	if strings.TrimSpace(c.apiKey) != "" {
		return strings.TrimSpace(c.apiKey)
	}
	return ChatGPTAPIKey()
}

func (c *ChatGPTClient) request(method, path string, body any, timeout time.Duration) ([]byte, error) {
	key := c.key()
	if key == "" {
		return nil, chatgptErrorf("no OpenAI API key: set OPENAI_API_KEY (or OPENAI_API_KEY_FILE) in the " +
			"environment of the process that talks to ChatGPT")
	}
	var reader io.Reader
	if body != nil {
		raw, err := json.Marshal(body)
		if err != nil {
			return nil, chatgptErrorf("cannot encode the request for %s: %v", path, err)
		}
		reader = bytes.NewReader(raw)
	}
	if timeout <= 0 {
		timeout = c.Timeout
	}
	ctx, cancel := context.WithTimeout(context.Background(), timeout)
	defer cancel()
	req, err := http.NewRequestWithContext(ctx, method, c.URL+path, reader)
	if err != nil {
		return nil, chatgptErrorf("cannot reach the OpenAI API at %s: %v", c.URL, err)
	}
	req.Header.Set("Authorization", "Bearer "+key)
	req.Header.Set("Accept", "application/json")
	if body != nil {
		req.Header.Set("Content-Type", "application/json")
	}
	if org := strings.TrimSpace(os.Getenv("OPENAI_ORG_ID")); org != "" {
		req.Header.Set("OpenAI-Organization", org)
	}
	if project := strings.TrimSpace(os.Getenv("OPENAI_PROJECT_ID")); project != "" {
		req.Header.Set("OpenAI-Project", project)
	}
	resp, err := c.client.Do(req)
	if err != nil {
		return nil, chatgptErrorf("cannot reach the OpenAI API at %s: %v", c.URL, err)
	}
	defer resp.Body.Close()
	raw, err := io.ReadAll(io.LimitReader(resp.Body, 32<<20))
	if err != nil {
		return nil, chatgptErrorf("cannot read the OpenAI answer to %s: %v", path, err)
	}
	if resp.StatusCode >= 400 {
		return nil, c.httpError(method, path, resp.StatusCode, raw)
	}
	return raw, nil
}

func (c *ChatGPTClient) httpError(method, path string, status int, raw []byte) error {
	detail := strings.TrimSpace(string(raw))
	if len(detail) > 1000 {
		detail = detail[:1000]
	}
	code, param := "", ""
	var doc struct {
		Error struct {
			Message string `json:"message"`
			Code    string `json:"code"`
			Param   string `json:"param"`
		} `json:"error"`
	}
	if json.Unmarshal(raw, &doc) == nil && doc.Error.Message != "" {
		detail, code, param = doc.Error.Message, doc.Error.Code, doc.Error.Param
	}
	hint := ""
	switch status {
	case 401:
		hint = " (check OPENAI_API_KEY)"
	case 403:
		hint = " (the key may not be allowed to use this model)"
	case 404:
		hint = fmt.Sprintf(" (does the key have access to model %q?)", c.Model)
	case 429:
		hint = " (rate limit or exhausted quota)"
	}
	return &ChatGPTError{
		Message: fmt.Sprintf("OpenAI %s %s failed with HTTP %d: %s%s", method, path, status, detail, hint),
		Status:  status, Code: code, Param: param,
	}
}

// Models lists the models the key may use (GET /models), as {"name", "id", "owned_by", "created"}.
func (c *ChatGPTClient) Models() ([]map[string]any, error) {
	raw, err := c.request("GET", "/models", nil, 0)
	if err != nil {
		return nil, err
	}
	var doc struct {
		Data []map[string]any `json:"data"`
	}
	if err := json.Unmarshal(raw, &doc); err != nil {
		return nil, chatgptErrorf("the OpenAI API returned invalid JSON for /models: %v", err)
	}
	if doc.Data == nil {
		return nil, chatgptErrorf("unexpected /models response (no 'data' list)")
	}
	models := make([]map[string]any, 0, len(doc.Data))
	for _, item := range doc.Data {
		id, _ := item["id"].(string)
		if id == "" {
			continue
		}
		models = append(models, map[string]any{
			"name": id, "id": id, "owned_by": item["owned_by"], "created": item["created"],
		})
	}
	return models, nil
}

// Available reports whether a key is configured and the API answers.
func (c *ChatGPTClient) Available() bool {
	_, err := c.Models()
	return err == nil
}

// optionalFields are the body fields worth dropping and retrying when a model rejects them.
var optionalFields = []string{"temperature", "response_format"}

var (
	rejectedPattern = regexp.MustCompile(`(?i)unsupported|unrecognized|not supported|does not support|unknown parameter`)
	quotedName      = regexp.MustCompile(`['"` + "`" + `]([A-Za-z_][A-Za-z0-9_]*)['"` + "`" + `]`)
	argumentName    = regexp.MustCompile(`(?i)argument(?: supplied)?:\s*([A-Za-z_][A-Za-z0-9_]*)`)
)

// rejectedField names the optional body field an error blames, when dropping it is worth one retry.
func rejectedField(err error, body map[string]any) string {
	var failure *ChatGPTError
	if !errors.As(err, &failure) || failure.Status != 400 || !rejectedPattern.MatchString(failure.Message) {
		return ""
	}
	names := []string{}
	if failure.Param != "" {
		names = append(names, failure.Param)
	}
	for _, match := range argumentName.FindAllStringSubmatch(failure.Message, -1) {
		names = append(names, match[1])
	}
	for _, match := range quotedName.FindAllStringSubmatch(failure.Message, -1) {
		names = append(names, match[1])
	}
	for _, name := range names {
		if _, present := body[name]; !present {
			continue
		}
		for _, field := range optionalFields {
			if name == field {
				return name
			}
		}
	}
	return ""
}

// Generate runs one non-streaming completion (POST /chat/completions).  A model
// that rejects an optional field (the reasoning models refuse temperature, older
// ones response_format) is asked again without it instead of failing.
func (c *ChatGPTClient) Generate(prompt string, o LLMOptions) (string, error) {
	model := strings.TrimSpace(o.Model)
	if model == "" {
		model = c.Model
	}
	messages := []map[string]any{}
	if o.System != "" {
		messages = append(messages, map[string]any{"role": "system", "content": o.System})
	}
	messages = append(messages, map[string]any{"role": "user", "content": prompt})
	body := map[string]any{"model": model, "messages": messages}
	if o.Temperature > 0 {
		body["temperature"] = o.Temperature
	}
	if o.JSON {
		body["response_format"] = map[string]any{"type": "json_object"}
	}
	for {
		raw, err := c.request("POST", "/chat/completions", body, o.Timeout)
		if err != nil {
			if field := rejectedField(err, body); field != "" {
				delete(body, field) // this model does not take that field: ask again without it
				continue
			}
			return "", err
		}
		return chatCompletionText(raw)
	}
}

// chatCompletionText is the assistant text of a chat-completions response.
func chatCompletionText(raw []byte) (string, error) {
	var doc struct {
		Choices []struct {
			FinishReason string `json:"finish_reason"`
			Message      struct {
				Content any    `json:"content"`
				Refusal string `json:"refusal"`
			} `json:"message"`
		} `json:"choices"`
	}
	if err := json.Unmarshal(raw, &doc); err != nil {
		return "", chatgptErrorf("the OpenAI API returned invalid JSON for /chat/completions: %v", err)
	}
	if len(doc.Choices) == 0 {
		return "", chatgptErrorf("unexpected chat completion (no 'choices')")
	}
	choice := doc.Choices[0]
	switch content := choice.Message.Content.(type) {
	case string:
		if strings.TrimSpace(content) != "" {
			return content, nil
		}
	case []any: // content parts instead of one string
		var text strings.Builder
		for _, part := range content {
			if item, ok := part.(map[string]any); ok {
				if value, ok := item["text"].(string); ok {
					text.WriteString(value)
				}
			}
		}
		if strings.TrimSpace(text.String()) != "" {
			return text.String(), nil
		}
	}
	if strings.TrimSpace(choice.Message.Refusal) != "" {
		return "", chatgptErrorf("the model refused to answer: %s", strings.TrimSpace(choice.Message.Refusal))
	}
	return "", chatgptErrorf("the model returned an empty answer (finish_reason=%q)", choice.FinishReason)
}

File: RadixCyclicNN/radixnet/chatgpt.py
"""ChatGPT (OpenAI) as a tutor: the same client interface as Ollama, over the hosted API.

:class:`ChatGPTClient` speaks OpenAI's chat-completions API (HTTP, standard
library only) and offers exactly what :class:`~radixnet.ollama.OllamaClient`
offers — :meth:`~ChatGPTClient.models`, :meth:`~ChatGPTClient.available`,
:meth:`~ChatGPTClient.generate`, :meth:`~ChatGPTClient.chat` — so anything that
takes a client (the code-generation tutor and its judge above all) can be
pointed at ChatGPT instead of a local model without further changes.

Configuration comes from the environment, the way the OpenAI tools expect it:

* ``OPENAI_API_KEY`` — the key (or ``OPENAI_API_KEY_FILE``, a file holding it,
  for Docker / Compose secrets).  There is no default and no way to pass a key
  through the HTTP API: the server uses the key of the process it runs in.
* ``OPENAI_BASE_URL`` — the endpoint, ``https://api.openai.com/v1`` by default;
  point it at any OpenAI-compatible server (llama.cpp, vLLM, LM Studio, a
  gateway) to use that instead.
* ``RADIXNET_OPENAI_MODEL`` — the default model, else ``gpt-4o-mini``.
* ``OPENAI_ORG_ID`` / ``OPENAI_PROJECT_ID`` — optional account headers.

Ollama's ``options`` names are translated (``num_predict`` ->
``max_completion_tokens``, and so on), and a model that rejects an optional
field — the reasoning models refuse ``temperature``, older ones refuse
``response_format`` — is retried without it instead of failing, so the tutor
works across model families.

Unlike the local provider this sends every prompt (problem statements, the
network's own programs) to OpenAI and costs money per call.
"""

from __future__ import annotations

import json
import os
import re
import urllib.error
import urllib.parse
import urllib.request
from typing import Any

from .llm import LLMError

__all__ = [
    "DEFAULT_MODEL",
    "DEFAULT_TIMEOUT",
    "DEFAULT_URL",
    "ChatGPTClient",
    "ChatGPTError",
    "api_key",
    "api_key_configured",
    "normalise_url",
]

DEFAULT_TIMEOUT = 120.0
"""Seconds to wait for one ChatGPT answer."""

OPENAI_URL = "https://api.openai.com/v1"
"""The hosted API; ``OPENAI_BASE_URL`` overrides it (any OpenAI-compatible server)."""

_LOCAL_HOSTS = ("localhost", "127.0.0.1", "::1", "[::1]", "0.0.0.0", "host.docker.internal")


def _is_local(host: str) -> bool:
    """True for an OpenAI-compatible server on this machine (llama.cpp, vLLM, LM Studio, a test double)."""
    host = (host or "").lower()
    return host in _LOCAL_HOSTS or host.endswith(".local")


def normalise_url(url: str | None) -> str:
    """``host:port`` or a full URL -> a base URL without a trailing slash (``/v1`` added when there is no path).

    Refuses to carry a key over plain ``http://`` to anything but a local
    server unless ``RADIXNET_OPENAI_ALLOW_INSECURE`` is set.
    """
    text = (url or "").strip()
    if not text:
        raise ValueError("the OpenAI base URL is empty")
    if "://" not in text:
        text = "https://" + text
    parts = urllib.parse.urlsplit(text)
    if parts.scheme not in ("http", "https"):
        raise ValueError(f"the OpenAI base URL must be http:// or https:// (got {url!r})")
    if not parts.netloc:
        raise ValueError(f"the OpenAI base URL has no host (got {url!r})")
    host = (parts.hostname or "").lower()
    if parts.scheme == "http" and not _is_local(host) and not _flag("RADIXNET_OPENAI_ALLOW_INSECURE"):
        raise ValueError(
            f"refusing to send the API key unencrypted to {host!r}: use https://, a local server, "
            "or set RADIXNET_OPENAI_ALLOW_INSECURE=1"
        )
    return f"{parts.scheme}://{parts.netloc}{parts.path.rstrip('/') or '/v1'}"


def _flag(name: str) -> bool:
    return os.environ.get(name, "").strip().lower() in ("1", "true", "yes", "on")


def _default_url() -> str:
    configured = os.environ.get("OPENAI_BASE_URL", "").strip() or os.environ.get("OPENAI_API_BASE", "").strip()
    try:
        return normalise_url(configured) if configured else OPENAI_URL
    except ValueError:  # a broken environment must not stop the package from importing
        return OPENAI_URL


DEFAULT_URL = _default_url()
DEFAULT_MODEL = os.environ.get("RADIXNET_OPENAI_MODEL", "").strip() or "gpt-4o-mini"


def api_key(explicit: str | None = None) -> str | None:
    """The API key: ``explicit``, else ``$OPENAI_API_KEY``, else the content of ``$OPENAI_API_KEY_FILE``."""
    if explicit and explicit.strip():
        return explicit.strip()
    key = os.environ.get("OPENAI_API_KEY", "").strip()
    if key:
        return key
    path = os.environ.get("OPENAI_API_KEY_FILE", "").strip()
    if path:
        try:
            with open(path, encoding="utf-8") as fh:
                return fh.read().strip() or None
        except OSError:
            return None
    return None


def api_key_configured(explicit: str | None = None) -> bool:
    """True when a key is available (its value is never returned or logged)."""
    return bool(api_key(explicit))


class ChatGPTError(LLMError):
    """The OpenAI API is unreachable, refused the request, or answered something unusable."""

    def __init__(self, message: str, *, status: int | None = None, code: str | None = None, param: str | None = None) -> None:
        super().__init__(message)
        self.status = status
        self.code = code
        self.param = param


# ---------------------------------------------------------------------------
# request bodies
# ---------------------------------------------------------------------------

_OPTION_NAMES = {
    "temperature": "temperature",
    "top_p": "top_p",
    "seed": "seed",
    "stop": "stop",
    "presence_penalty": "presence_penalty",
    "frequency_penalty": "frequency_penalty",
    "num_predict": "max_completion_tokens",  # Ollama's name for the answer length
    "max_tokens": "max_completion_tokens",
    "max_completion_tokens": "max_completion_tokens",
}

_OPTIONAL_FIELDS = (
    "temperature", "top_p", "seed", "stop", "presence_penalty", "frequency_penalty",
    "max_completion_tokens", "response_format",
)
"""Body fields worth dropping and retrying when a model rejects them."""

_REJECTED = re.compile(r"unsupported|unrecognized|not supported|does not support|unknown parameter", re.IGNORECASE)
_QUOTED_NAME = re.compile(r"['\"`]([A-Za-z_][A-Za-z0-9_]*)['\"`]")
_ARGUMENT_NAME = re.compile(r"argument(?: supplied)?:\s*([A-Za-z_][A-Za-z0-9_]*)", re.IGNORECASE)


def _options_body(options: dict | None) -> dict:
    """Ollama-style ``options`` -> chat-completions fields (unknown names are dropped)."""
    body: dict[str, Any] = {}
    for name, value in (options or {}).items():
        field = _OPTION_NAMES.get(name)
        if field is None or value is None:
            continue
        if field == "max_completion_tokens":
            try:
                value = int(value)
            except (TypeError, ValueError):
                continue
            if value <= 0:  # Ollama's "no limit"
                continue
        body[field] = value
    return body


def _rejected_field(error: ChatGPTError, body: dict) -> str | None:
    """The optional body field an error blames, when dropping it is worth one retry."""
    if error.status != 400 or not _REJECTED.search(str(error)):
        return None
    message = str(error)
    names = [error.param] if isinstance(error.param, str) else []
    names += _ARGUMENT_NAME.findall(message)
    names += _QUOTED_NAME.findall(message)
    for name in names:
        if name in _OPTIONAL_FIELDS and name in body:
            return name
    return None


def _answer(data: Any) -> str:
    """The assistant text of a chat-completions response."""
    choices = data.get("choices") if isinstance(data, dict) else None
    if not isinstance(choices, list) or not choices or not isinstance(choices[0], dict):
        raise ChatGPTError("unexpected chat completion (no 'choices')")
    choice = choices[0]
    message = choice.get("message") if isinstance(choice.get("message"), dict) else {}
    content = message.get("content")
    if isinstance(content, list):  # content parts instead of one string
        content = "".join(part.get("text", "") for part in content if isinstance(part, dict))
    if isinstance(content, str) and content.strip():
        return content
    refusal = message.get("refusal")
    if isinstance(refusal, str) and refusal.strip():
        raise ChatGPTError(f"the model refused to answer: {refusal.strip()}")
    raise ChatGPTError(f"the model returned an empty answer (finish_reason={choice.get('finish_reason')!r})")


# ---------------------------------------------------------------------------
# the client
# ---------------------------------------------------------------------------


class ChatGPTClient:
    """Minimal client for OpenAI's API (``GET /models``, ``POST /chat/completions``).

    ``url``, ``model`` and ``timeout`` fall back to the module defaults; the
    key comes from ``api_key`` or the environment and is read at request time,
    never stored in a report, a job record or a log line.
    """

    provider = "chatgpt"

    __slots__ = ("url", "model", "timeout", "_api_key", "_local", "organization", "project")

    def __init__(
        self,
        url: str | None = None,
        model: str | None = None,
        timeout: float | None = None,
        *,
        api_key: str | None = None,
    ) -> None:
        self.url = normalise_url(url or DEFAULT_URL)
        self._local = _is_local(urllib.parse.urlsplit(self.url).hostname or "")
        self.model = (model or DEFAULT_MODEL).strip() or DEFAULT_MODEL
        self.timeout = float(timeout) if timeout else DEFAULT_TIMEOUT
        self._api_key = api_key.strip() if api_key and api_key.strip() else None
        self.organization = os.environ.get("OPENAI_ORG_ID", "").strip() or None
        self.project = os.environ.get("OPENAI_PROJECT_ID", "").strip() or None

    def __repr__(self) -> str:  # never shows the key
        return f"ChatGPTClient(url={self.url!r}, model={self.model!r})"

    @property
    def configured(self) -> bool:
        """True when a key is available for this client."""
        return api_key_configured(self._api_key)

    # -- transport -----------------------------------------------------------

    def _key(self) -> str:
        key = api_key(self._api_key)
        if not key:
            raise ChatGPTError(
                "no OpenAI API key: set OPENAI_API_KEY (or OPENAI_API_KEY_FILE) in the environment of the "
                "process that talks to ChatGPT"
            )
        return key

    def _headers(self, body: bool) -> dict[str, str]:
        headers = {"Authorization": f"Bearer {self._key()}", "Accept": "application/json"}
        if body:
            headers["Content-Type"] = "application/json"
        if self.organization:
            headers["OpenAI-Organization"] = self.organization
        if self.project:
            headers["OpenAI-Project"] = self.project
        return headers

    def _http_error(self, exc: urllib.error.HTTPError, path: str) -> ChatGPTError:
        raw = exc.read().decode("utf-8", errors="replace")[:1000].strip()
        detail, code, param = raw, None, None
        try:
            error = json.loads(raw).get("error")
        except (ValueError, AttributeError):
            error = None
        if isinstance(error, dict):
            detail = str(error.get("message") or raw)
            code = error.get("code") if isinstance(error.get("code"), str) else None
            param = error.get("param") if isinstance(error.get("param"), str) else None
        elif isinstance(error, str):
            detail = error
        hint = {
            401: " (check OPENAI_API_KEY)",
            403: " (the key may not be allowed to use this model)",
            404: f" (does the key have access to model {self.model!r}?)",
            429: " (rate limit or exhausted quota)",
        }.get(exc.code, "")
        return ChatGPTError(
            f"OpenAI {path} failed with HTTP {exc.code}: {detail or exc.reason}{hint}",
            status=exc.code, code=code, param=param,
        )

    def _request(self, method: str, path: str, body: dict | None = None, timeout: float | None = None) -> Any:
        data = None if body is None else json.dumps(body).encode("utf-8")
        request = urllib.request.Request(self.url + path, data=data, method=method, headers=self._headers(data is not None))
        # The hosted API goes through the environment's proxy; a server on this machine never does.
        opener = urllib.request.build_opener(urllib.request.ProxyHandler({})) if self._local else urllib.request.build_opener()
        try:
            with opener.open(request, timeout=timeout or self.timeout) as response:
                raw = response.read()
        except urllib.error.HTTPError as exc:
            raise self._http_error(exc, f"{method} {path}") from exc
        except (urllib.error.URLError, OSError, TimeoutError) as exc:
            reason = getattr(exc, "reason", None) or exc
            raise ChatGPTError(f"cannot reach the OpenAI API at {self.url}: {reason}") from exc
        try:
            return json.loads(raw.decode("utf-8"))
        except ValueError as exc:
            raise ChatGPTError(f"the OpenAI API returned invalid JSON for {path}: {exc}") from exc

    # -- API -----------------------------------------------------------------

    def models(self) -> list[dict]:
        """Models the key can use (``GET /models``), as ``{"name", "id", "owned_by", "created"}`` sorted by name."""
        data = self._request("GET", "/models")
        items = data.get("data") if isinstance(data, dict) else None
        if not isinstance(items, list):
            raise ChatGPTError("unexpected /models response (no 'data' list)")
        models = [
            {"name": str(m.get("id")), "id": str(m.get("id")), "owned_by": m.get("owned_by"), "created": m.get("created")}
            for m in items
            if isinstance(m, dict) and m.get("id")
        ]
        return sorted(models, key=lambda m: m["name"])

    def available(self) -> bool:
        """True when a key is configured and the API answers (never raises)."""
        try:
            self.models()
        except LLMError:
            return False
        return True

    def _complete(
        self, messages: list[dict], model: str | None, json_mode: bool, options: dict | None, timeout: float | None
    ) -> str:
        body: dict[str, Any] = {"model": model or self.model, "messages": messages}
        body.update(_options_body(options))
        if json_mode:
            body["response_format"] = {"type": "json_object"}
        while True:
            try:
                return _answer(self._request("POST", "/chat/completions", body, timeout))
            except ChatGPTError as exc:
                rejected = _rejected_field(exc, body)
                if rejected is None:
                    raise
                body.pop(rejected)  # this model does not take that field: ask again without it

    def generate(
        self,
        prompt: str,
        *,
        system: str | None = None,
        model: str | None = None,
        json_mode: bool = False,
        options: dict | None = None,
        timeout: float | None = None,
    ) -> str:
        """One completion (``POST /chat/completions`` with a single user turn); ``json_mode`` asks for JSON."""
        messages: list[dict] = []
        if system:
            messages.append({"role": "system", "content": system})
        messages.append({"role": "user", "content": prompt})
        return self._complete(messages, model, json_mode, options, timeout)

    def chat(
        self,
        messages: list[dict],
        *,
        model: str | None = None,
        json_mode: bool = False,
        options: dict | None = None,
        timeout: float | None = None,
    ) -> str:
        """One chat turn; ``messages`` are ``{"role", "content"}`` dicts."""
        return self._complete([dict(m) for m in messages], model, json_mode, options, timeout)
